Stm32h7 flash support

nicovince opened the pull request:
Added support to read/write flash on stm32h743.

Tested on a nucleo_h743zi with samples/drivers/flash_shell.

Fixes #29831

explora26 commented:
Please notice that MCUBoot is not support due to the minimum flash write size over 8 bytes.
https://github.com/mcu-tools/mcuboot/issues/713

nicovince commented:
> Please notice that MCUBoot is not support due to the minimum flash write size over 8 bytes.
> [mcu-tools/mcuboot#713](https://github.com/mcu-tools/mcuboot/issues/713)

Indeed, after publishing my PR, I tried to play with MCUBoot, and noticed that the swap would fail. As far as I could test, I did not notice any issues while playing around with flash_shell sample.

erwango commented on drivers/flash/flash_stm32.c:
```diff
@@ -141,6 +144,9 @@ int flash_stm32_wait_flash_idle(const struct device *dev)
 #if defined(CONFIG_SOC_SERIES_STM32G0X)
 	while ((FLASH_STM32_REGS(dev)->SR & FLASH_SR_BSY1)) {
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
+	while ((FLASH_STM32_REGS(dev)->SR1 & (FLASH_SR_BSY | FLASH_SR_QW))
+	       || (FLASH_STM32_REGS(dev)->SR2 & (FLASH_SR_BSY | FLASH_SR_QW))) {
```
Only checking the QW bit should be sufficient.
Though, you should check that _both_ banks are not busy (but only under #if DUAL_BANK).

erwango replied:
I'm "unresolving" this until I can see it has been actually resolved.

erwango commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -0,0 +1,219 @@
+#define LOG_LEVEL CONFIG_FLASH_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_DOMAIN);
+#define STM32H7X_SECTOR_SHIFT 17
```
Some H7 references use 8K sectors.

nicovince replied:
```diff
-#define STM32H7X_SECTOR_SHIFT 17
+
+#define LOG2CEIL(x) ((((x) <= 2048) ? 11 : (((x) <= 4096)?12:(((x) <= 8192) ? \
+       13 : (((x) <= 16384) ? 14 : (((x) <= 32768) ? 15:(((x) <= 65536) ? \
+       16 : (((x) <= 131072) ? 17 : (((x) <= 262144) ? 18:(((x) <= 524288) ? \
+       19 : (((x) <= 1048576) ? 20 : (((x) <= 2097152) ? \
+       21 : (((x) <= 4194304) ? 22 : (((x) <= 8388608) ? \
+       23 : (((x) <= 16777216) ? 24 : (((x) <= 33554432) ? \
+       25 : (((x) <= 67108864) ? 26 : (((x) <= 134217728) ? \
+       27 : (((x) <= 268435456) ? 28 : (((x) <= 536870912) ? \
+       29 : (((x) <= 1073741824) ? 30 : (((x) <= 2147483648) ? \
+       31 : 32))))))))))))))))))))))
+#define STM32H7X_SECTOR_SHIFT LOG2CEIL(FLASH_BANK_SIZE / FLASH_SECTOR_TOTAL)
```
What do you think about this ? I copied the macro LOG2CEIL from `include/linker/linker-tool-mwdt.h`
This way the SECTOR_SHIFT is computed automatically for every devices in the family.

erwango replied:
What about simplifying a bit things and use the following in `get_sector` function:
```
static unsigned int get_sector(off_t offset)
{
	return offset / FLASH_SECTOR_SIZE;
}
```

erwango commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -0,0 +1,219 @@
+	uint32_t tmp;
+	volatile uint32_t *cr;
+
+	if (sector <= 7) {
```
For 8K sectors, Bank2 starts at sector 128.

erwango replied:
This would work indeed

erwango commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -0,0 +1,219 @@
+
+	if (sector <= 7) {
+		cr = &regs->CR1;
+	} else {
+		cr = &regs->CR2;
+	}
```
Only if DUAL_BANK

erwango commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -0,0 +1,219 @@
+	return 0;
+}
+
+static int write_4dwords(const struct device *dev,
```
Some H7 references will use 128 bits programming length

nicovince replied:
Rewriting the function to write_ndwords and taking an extra parameters n to define the number of double words to be written

erwango reviewed (changes requested):
Thanks for this contribution! STM32H7 is not the easiest piece of the family and its flash controller block is no exception.

I'm affraid there are numerous changes to be addressed for driver completion.
Proposed driver does not takes sufficiently into account:
- Series heterogeneity (Single vs Dual Bank, number of sectors, size of sectors, programming size, ..)
- Dual Bank simultaneous operation support

Considering these points, for more flexibility in implementation and final readability, I would advise to create a dedicated stm32h7 driver (ie not based on flash_stm32.c).
Though, I don't request to implement the full series support with all its variants (not all of these are supported today in zephyr). But at least the driver should be able to detect if the H7 variant is supported or not.

nicovince commented on drivers/flash/flash_stm32.c:
```diff
@@ -274,6 +280,26 @@ static int flash_stm32_write_protection(const struct device *dev, bool enable)
+			regs->KEYR1 = FLASH_KEY2;
+		}
+	}
+    /* Bank 2 */
+	if (enable) {
+		regs->CR2 |= FLASH_CR_LOCK;
+	} else {
+		if (regs->CR2 & FLASH_CR_LOCK) {
+			regs->KEYR2 = FLASH_KEY1;
+			regs->KEYR2 = FLASH_KEY2;
+		}
+	}
```
This is due to the API 
```c
typedef int (*flash_api_write_protection)(const struct device *dev,
                                          bool enable);
```
There are no parameters indicating on which part of the flash the write protection should occurs, therefore I lock both banks and lose the ability to operate on both banks.
I did not see how I could address this without changing the API.
Do you have any suggestions ?

erwango replied:
Right.
One alternative would be to use 2 different devices, but I'm ok to limit the driver to sequential usage of banks as a first step.

nicovince commented on drivers/flash/flash_stm32.c:
```diff
@@ -52,6 +52,9 @@ LOG_MODULE_REGISTER(flash_stm32, CONFIG_FLASH_LOG_LEVEL);
 /* STM32G4: maximum erase time of 24.47ms for a 2K sector */
 #elif defined(CONFIG_SOC_SERIES_STM32G4X)
 #define STM32_FLASH_MAX_ERASE_TIME	25
+/* TODO:STM32H7: maximum erase time of XXX ms for a 128K sector */
```
Why removing the `/*` and `*/` ?

erwango replied:
Copy/paste issue. These should be kept indeed

nicovince commented:
Hi, thanks for your thorough review.

> I would advise to create a dedicated stm32h7 driver (ie not based on flash_stm32.c).

I thought about it initially but I was afraid to have a lot of copy/paste from flash_stm32.c, so I decided to go with with the approach used for flash_stm32wbx.c: using ifdef to remove functions definitions from flash_stm32.c and defining them into flash_stm32wbx.c.
The downside is the multiple ifdef popping arround in flash_stm32.c
If that is not acceptable I'll see how i can improve it, maybe splitting flash_stm32.c into flash_stm32_cmn.c and flash_stm32_v2.c ? (there is already a v1).

Otherwise I've addressed the various suggestions you've made, and hopefully fixed the buildkite/zephyr.

I noticed an issue when using samples/drivers/flash_shell/, if i try to `flash write_pattern 0x20000 256` twice, on next reset I see the content which has been erased but the pattern is not there anymore. If you have any idea let me know.

Do you want me to update the PR with the commits I made to address your suggestions or shall I wait to have the issue above fixed ?

Thanks

erwango commented:
> I thought about it initially but I was afraid to have a lot of copy/paste from flash_stm32.c, so I decided to go with with the approach used for flash_stm32wbx.c: using ifdef to remove functions definitions from flash_stm32.c and defining them into flash_stm32wbx.c.
The downside is the multiple ifdef popping arround in flash_stm32.c
If that is not acceptable I'll see how i can improve it, maybe splitting flash_stm32.c into flash_stm32_cmn.c and flash_stm32_v2.c ? (there is already a v1).

If we consider the fact that H7 flash driver might require in the end to support simultaneous operations to 2 banks, I think that we'll end up hacking a lot flash_stm32.c file and I'd prefer to avoid this situation. 
Also, I don't think splitting flash_stm32.c would be that great (flash_stm32_v1.c is used together with flash_stm32.c for F0, F1, F3 and L0), it is a different split as you're proposing for v2.
So in the end, the simplest is to opt for a monolithic flash_stm32_h7.c (this might require some file renaming to minimize confusion with other stm32 flash dirvers, but I'll take care of that part).

nicovince commented:
@erwango, I have addressed the various suggestions you've made, and properly invalidated data cache on erased sectors.

Please, let me know if I have missed something.

Thanks.

erwango commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -0,0 +1,534 @@
+			       size_t len)
+{
+	int rc;
+	off_t flush_offset = get_sector(offset) * FLASH_SECTOR_SIZE;
```
I'm a bit puzzled by this line given that get_sector(offset) = offset / FLASH_SECTOR_SIZE.
A subtlety I didn't catch maybe?

nicovince replied:
I noticed an issue when playing around with `zephyr/samples/drivers/flash_shell/`:

If i write a few bytes at the beginning of a sector, then do a `flash write_pattern` at an address further away in that sector, I would still see the few bytes written at the beginning of the sector even though `write_pattern` performs an erase on the sector.

So, with this line I'm telling the driver I want to invalidate the cache at the first address of the sector which has been erased.

erwango replied:
Ok, I'll check that

erwango commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -0,0 +1,534 @@
+{
+	int rc;
+	off_t flush_offset = get_sector(offset) * FLASH_SECTOR_SIZE;
+	size_t flush_len = (((get_sector(offset + len) + 1) * FLASH_SECTOR_SIZE)
```
I'm also confused here. Why length to be used in SCB_InvalidateDCache_by_Addr would be different from len?

nicovince replied:
Same issue as above.
Although I suspect there is still a bug lying somewhere because if I do a write_pattern with a length of 256 at offset 0, then a write_pattern of length 32 at offset 0, I still see the data at offet 32 to 63.

erwango replied:
Ok, I saw that as well.
I would tend to say this is related to a bug in upper layers since this cannot be seen if `flash erase` is run manually between two `flash write_pattern`

erwango commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -0,0 +1,534 @@
+	uint8_t unaligned_datas[nbytes];
+	int suboffset = (offset % nbytes);
+
+	if (suboffset != 0) {
+		/* Offset is not aligned on FLASHWORD boundary */
+		offset -= suboffset;
+		memset(unaligned_datas, 0xff, sizeof(unaligned_datas));
+		for (i = 0; i < len && i < nbytes; ++i) {
+			unaligned_datas[suboffset + i] = ((uint8_t *)data)[i];
+		}
+		rc = write_ndwords(dev, offset,
+				   (const uint64_t *)unaligned_datas,
+				   ndwords);
+		i = 32;
+		offset += 32;
+	}
```
I don't think this is something we'd like to do.
If write request offset is not aligned on flash blocks boundaries we should reject and warn user, as done in other drivers.

erwango reviewed (changes requested):
Thanks for this new version, this is going in the good direction.
This seems overall fine, except the comments or clarifications required bellow.
One point I'm not sure is that if we should consider writing unaligned blocks. I think as a first version it would be better to reject as done on other drivers, unless there is a good reason to do so.
Btw, it will be required to remove the now unrelated first commits and start wit the commit that add the new driver

nicovince commented:
> One point I'm not sure is that if we should consider writing unaligned blocks. I think as a first version it would be better to reject as done on other drivers, unless there is a good reason to do so.
Since the flash words are pretty larges (32 bytes), I thought it was relevant, I don't mind removing it.

erwango commented:
> Since the flash words are pretty larges (32 bytes), I thought it was relevant, I don't mind removing it.

This is a fair point, but I'm not sure that upper layers are able to deal with this, While I'm sure they're able to deal with various write bloc sizes (this is why we provide a `flash_parameters.write_block_size `). Anyway, if this appear to be possible and requested, it could be added later. Let's start with a safe implementation.

nicovince commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -0,0 +1,534 @@
+		return -EIO;
+	}
+#ifdef DUAL_BANK
+	while ((FLASH_STM32_REGS(dev)->SR1 & FLASH_SR_QW)
+	       || (FLASH_STM32_REGS(dev)->SR2 & FLASH_SR_QW))
```
Are you sure ?
I want to wait while QW bit is set in either of the bank, not both of them simultaneously.

erwango replied:
My bad, this is fine indeed.

nicovince commented:
@erwango, I have addressed your suggested changes, except for [this one](https://github.com/zephyrproject-rtos/zephyr/pull/30043#discussion_r530468924).

Thanks for your review.

erwango reviewed (changes requested):
Thanks for the update, few more comments

erwango commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -0,0 +1,528 @@
+			LOG_ERR("Write offset not aligned on flashword length. "
+				"Offset: 0x%x, flashword length: %d",
+				offset, FLASH_NB_32BITWORD_IN_FLASHWORD * 4);
+			return -EINVAL;
```
Function is bool.

erwango reviewed (changes requested):
Final comments on my side.
Would you mind :
- refactor the PR in order to have only one commit for the driver (no fix commts...)
- add the minimum flash descriptions in dtsi files for available H7 SoCs
- rebase on master in order to solve the merge conflict

Txs!

nicovince commented:
>     * add the minimum flash descriptions in dtsi files for available H7 SoCs

I'm not sure to understand exactly what that means. Is it related to the write-block-size/erase-block-size properties ?

erwango commented:
> I'm not sure to understand exactly what that means. Is it related to the write-block-size/erase-block-size properties ?

Yes!

erwango commented on boards/arm/nucleo_h743zi/nucleo_h743zi.dts:
```diff
@@ -112,7 +113,52 @@
 		     &eth_txd1_pb13>;
 };
 
+
```
Remove this line

erwango commented on boards/arm/nucleo_h743zi/nucleo_h743zi.dts:
```diff
@@ -112,7 +113,52 @@
+&flash0 {
+	/*
+	 * For more information, see:
+	 * https://docs.zephyrproject.org/latest/reference/devicetree/index.html#fixed-flash-partitions
```
This link is obsolete (I should add again, as this is at least the 2nd time the link used for this doc changes).
Let's remove the whole comment as this is now a common feature and there is no specific reason to comment it more than other dts pieces.
We might need to clean this up from other boards as well.

erwango reviewed (changes requested):
Sorry if that was not clear but "one commit for the driver" and ... other commits can be kept :-)
Anyway, now that everything is squashed, it will be easy to split into:
- 1 dts/ commit
- 1 soc/ commit
- 1 drivers/ commit
     -> For this one, please rename to something like drivers/flash: Add support for ....
         and add the "Fixes #29831" as part of the commit message. This way, related issue will be closed automatically once PR is 
         merged
- 1 board commit

Txs!

nvincent-vossloh committed:
dts: flash: write/erase block size to stm32h7 devs

Define write and erase block size for supported stm32h7 devices
Use a specific compatible string for stm32h7 devices for the flash
driver

Signed-off-by: Nicolas VINCENT <nicolas.vincent@vossloh.com>

nvincent-vossloh committed:
drivers/flash: add support for stm32h7 devices

Fixes #29831: Implements flash driver for stm32h7 devices.
The driver is independant from the other stm32 families (flash_stm32.c),
only the header interface is (mainly) common.

Signed-off-by: Nicolas VINCENT <nicolas.vincent@vossloh.com>

nvincent-vossloh committed:
soc: stm32h7: define rom offset when using mcuboot

Set rom offset to 0x400 if application is compiled with
CONFIG_BOOTLOADER_MCUBOOT.
Please note that mcuboot is not yet supported on stm32h7 devices

Signed-off-by: Nicolas VINCENT <nicolas.vincent@vossloh.com>

nvincent-vossloh committed:
boards: define flash partitions for nucleo_h743zi

Defines partitions that can be used by mcuboot on nucleo_h743zi board.
Please note that mcuboot is not yet supported on stm32 h7 family as the
write-block-size is greater than 8.

Signed-off-by: Nicolas VINCENT <nicolas.vincent@vossloh.com>

nicovince commented:
Latest push should have fixed the buildkite/zephyr. Sorry about that...

ABOSTM commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -0,0 +1,531 @@
+	size_t flush_len = (((get_sector(offset + len) + 1) * FLASH_SECTOR_SIZE)
+			    - flush_offset);
+
+	if (!flash_stm32_valid_range(dev, offset, len, false)) {
```
From my point of view erase is a kind of write (we write 0xFF to a location) and thus last parameter of flash_stm32_valid_range() should be true. Write is opposed to read (and not opposed to erase).

ABOSTM commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -0,0 +1,531 @@
+static const struct flash_pages_layout stm32h7_flash_layout[] = {
+	{ .pages_count = FLASH_SECTOR_TOTAL, .pages_size = FLASH_SECTOR_SIZE },
+#ifdef DUAL_BANK
+	{ .pages_count = FLASH_SECTOR_TOTAL, .pages_size = FLASH_SECTOR_SIZE },
```
I think you will face the same issue I have on STM32G4 #30231
There may be some discontinuity between bank1 and bank2 on some chip, when dual bank, with smaller flash .
Ex: stm32h747zi has 2M flash but stm32h747IG has only 1M flash, in this case real number of sector will be smaller, and by the way you cannot rely on FLASH_SECTOR_SIZE which is defined to 8 for all soc STM32H747xx.

nicovince replied:
I've had a look at files under `modules/hal/stm32/stm32cube/stm32h7xx/soc`, and all SOCs defining `DUAL_BANK` have a 1MB `FLASH_BANK_SIZE` (based on the comments).
So as far as I understand it the h7 family would not be impacted by this fragmented issue. What do you think ?

ABOSTM replied:
Unfortunately, it seems thoses files are erroneous. As I said, stm32h747IG has only 1M flash (512 HB per bank).
I submitted an internal ticket to change FLASH_SIZE definition.
And so h7 family is impacted by potential discontinuity.
As FLASH_SIZE (as well as FLASH_BANK_SIZE) is not reliable for now, you can use `LL_GetFlashSize()`

ABOSTM replied:
@nicovince, after discussion with @erwango,
we propose you to not implement discontinuity management in this PR,
and I will do it in a separate PR.

nicovince replied:
Ok, my bad.

I'll look into it then. How can I test my code on this particular chip ?

ABOSTM reviewed (approved):
LGTM

erwango commented:
@nicovince would you moind rebasing to fix CI issue ?

nicovince commented:
> @nicovince would you mind rebasing to fix CI issue ?

Done.

nvlsianpu commented:
@nicovince reopen?

nicovince commented:
> @nicovince reopen?

I'm still trying to figure out why the CI built successfully when I finalized this PR and why it failed later...
I'm pretty new to zephyr development, what shoud i do ? 
Repoen #29831 and create a new PR, or continue on this one ?

nvlsianpu commented:
Reopen this time means submit another PR.

erwango commented:
> I'm still trying to figure out why the CI built successfully when I finalized this PR and why it failed later...

@nicovince, there is a trick indeed.
My understanding of what happened is the following: 
In CI, only a limited subset of platforms vs samples/tests are build. On top of that a specific run is launched for each impacted board (based on path board/foor/bar).
On daily builds, all test/samples vs platforms combinations are run. And I guess this is where the _m4 targets builds were launched.

By the way, I'm looking at the issue.
Definition of flash controller on H7 M4 cores have some consequences that should be checked.
I'm not sure exactly how this should be fixed.
I propose to push a new PR with your commits when I'm done

nicovince commented:
@erwango, I was getting to the same conclusion.

I started modifying the dtsi in order to take into account the different flash address based on the core (0x800_0000 for m7, 0x810_0000 for the m4), and the dts compilation passes, but another failure rises at linking stage regarding the cache invalidation call (SCB_).
Setting that call under ifdnef CORE_CM4 does the trick...
Is there an invalidate cache call on M4 cores ?

erwango commented:
> Is there an invalidate cache call on M4 cores ?

No. No flash cache mgmt on H7 M4 cores it seems. 
Anyway, there are few more hickups on the M4 core flash driver support.
So I ended up thinking that flash driver support on M4 core should be disabled cleanly for now on.

GiancarloPerroni commented:
Hi, I ve some questions:

- Is for M4 core a flash-drviver support foreseen? When?
- what is necessary to do in order to adapt M7 driver to make it running with M4?

Thank you in advance for any answer.
